kratos/factories/builder_and_solver_factory.h

```diff
@@ -0,0 +1,165 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Vicente Mataix Ferrandiz
+//
+
+#if !defined(KRATOS_BUILDER_AND_SOLVER_FACTORY_H_INCLUDED )
+#define  KRATOS_BUILDER_AND_SOLVER_FACTORY_H_INCLUDED
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "includes/kratos_parameters.h"
+#include "includes/kratos_components.h"
+#include "includes/shared_pointers.h"
+#include "linear_solvers/linear_solver.h"
+#include "solving_strategies/builder_and_solvers/builder_and_solver.h"
+#include "spaces/ublas_space.h"
+
+namespace Kratos
+{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/**
+ * @class BuilderAndSolverFactory
+ * @ingroup KratosCore
+ * @brief Here we add the functions needed for the registration of builder and solvers
+ * @details Defines the base builder and solver factory
+ * @author Vicente Mataix Ferrandiz
+ * @tparam TSparseSpace The sparse space definition
+ * @tparam TLocalSpace The dense space definition
+ * @tparam TLinearSolver The linear solver type considered
+ */
+template< typename TSparseSpace, typename TLocalSpace, class TLinearSolver>
+class BuilderAndSolverFactory
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// The definition of the custom class
+    typedef BuilderAndSolverFactory<TSparseSpace,TLocalSpace, TLinearSolver> FactoryType;
+
+    // The definition of the linear solver type
+    typedef LinearSolver<TSparseSpace,TLocalSpace> LinearSolverType;
+
+    /// The definition of the builder and solver
+    typedef BuilderAndSolver<TSparseSpace,TLocalSpace, TLinearSolver> BuilderAndSolverType;
+
+    /// Pointer definition of BuilderAndSolverFactory
+    KRATOS_CLASS_POINTER_DEFINITION(BuilderAndSolverFactory );
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief This method checks if the linear builder and solver is registered
+     * @return True if registered, false otherwise
+     */
+    virtual bool Has(const std::string& rSolverType)
+    {
+        return KratosComponents< FactoryType >::Has( rSolverType );
+    }
+
+    /**
+     * @brief This method creates a new builder and solver
+     * @return The pointer to the builder and solver of interest
+     */
+    virtual typename BuilderAndSolverType::Pointer Create(typename LinearSolverType::Pointer pLinearSolver, Kratos::Parameters Settings)
+    {
+        const std::string& builder_and_solver_type = Settings["builder_and_solver_type"].GetString();
```

Comment:
here the check name is builder_and_solver_type

Reply:
I don't agree:
```c++
template<class TBaseClass> 
TBaseClass::Pointer Create(Parameters){
const std::string& object_type = Settings[TBaseClass::Type()].GetString();
}
```

Reply:
OK, I will see later. The last changes I pushed were basically to solve the issues with the compilation and conflicts with the master. Not a final solution, in fact I still have an issue that I need to consult with @roigcarlo